modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Logic/Erc1155BalanceFetcher.swift

```diff
@@ -11,27 +11,44 @@ import AlphaWalletWeb3
 ///B. Fetch balance for each tokenId owned (now or previously. For the latter value would be 0)
 ///
 ///This class performs (B)
-public class Erc1155BalanceFetcher {
+final class Erc1155BalanceFetcher {
     private let address: AlphaWallet.Address
     private let server: RPCServer
+    private var inFlightPromises: [Int: Promise<[BigInt: BigUInt]>] = [:]
+    private let queue = DispatchQueue(label: "org.alphawallet.swift.erc1155BalanceFetcher")
 
-    public init(address: AlphaWallet.Address, server: RPCServer) {
+    init(address: AlphaWallet.Address, server: RPCServer) {
         self.address = address
         self.server = server
     }
 
-    public func fetch(contract: AlphaWallet.Address, tokenIds: Set<BigInt>) -> Promise<[BigInt: BigUInt]> {
-        //tokenIds must be unique (hence arg is a Set) so `Dictionary(uniqueKeysWithValues:)` wouldn't crash
-        let tokenIds = Array(tokenIds)
-        let address = EthereumAddress(self.address.eip55String)!
-        let addresses: [EthereumAddress] = [EthereumAddress](repeating: address, count: tokenIds.count)
-        return firstly {
-            callSmartContract(withServer: server, contract: contract, functionName: "balanceOfBatch", abiString: AlphaWallet.Ethereum.ABI.erc1155String, parameters: [addresses, tokenIds] as [AnyObject])
-        }.map { result in
-            guard let balances = result["0"] as? [BigUInt], balances.count == tokenIds.count else {
-                throw CastError(actualValue: result["0"], expectedType: [BigUInt].self)
+    func fetch(contract: AlphaWallet.Address, tokenIds: Set<BigInt>) -> Promise<[BigInt: BigUInt]> {
+        firstly {
+            .value(contract)
+        }.then(on: queue, { [weak self, queue, address, server] contract -> Promise<[BigInt: BigUInt]> in
+            let key = "\(contract.eip55String)-\(tokenIds.hashValue)".hashValue
```

Comment:
What's the reason to use the hash as the key instead of the original string? I thought there's a few reasons not to:

* hashing a string key is already done by the dictionary anyway and the keys are short (so they hash fast)
* string keys are more inspectable/debuggable

Reply:
yes, i agree with you, made it as `int` because it returns same length, so it might be optimized